apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1.IntegrationTests/BucketPolicyOnlyTest.cs

```diff
@@ -0,0 +1,129 @@
+﻿// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Google.Apis.Storage.v1.Data;
+using Google.Cloud.ClientTesting;
+using System;
+using System.IO;
+using System.Net;
+using Xunit;
+
+namespace Google.Cloud.Storage.V1.IntegrationTests
+{
+    [Collection(nameof(StorageFixture))]
+    public class BucketPolicyOnlyTest
+    {
+        private readonly StorageFixture _fixture;
+
+        public BucketPolicyOnlyTest(StorageFixture fixture) => _fixture = fixture;
+
+        [Fact]
+        public void EnableDisableOnExistingBucket()
+        {
+            string bucketName = _fixture.GenerateBucketName();
+            var client = _fixture.Client;
+            CreateBucket();
+            EnableBpo();
+            UploadObjectSucceeds_DownloadObjectFails();
+            DisableBpo();
+
+            // Individual steps of the test as local methods
+
+            void CreateBucket()
+            {
+                Bucket bucket = _fixture.CreateBucket(bucketName, false);
+                Assert.False(bucket.IamConfiguration.BucketPolicyOnly.Enabled);
+            }
+
+            void EnableBpo()
+            {
+                client.PatchBucket(CreateBucketRepresentation(bucketName, true));
+
+                var fetched = client.GetBucket(bucketName);
+                Assert.True(fetched.IamConfiguration.BucketPolicyOnly.Enabled);
+                var lockedTime = fetched.IamConfiguration.BucketPolicyOnly.LockedTime;
+                Assert.NotNull(lockedTime);
+
+                var expectedLockedTime = DateTime.UtcNow.AddDays(90);
+                Assert.InRange(lockedTime.Value.ToUniversalTime(), expectedLockedTime.AddMinutes(-10), expectedLockedTime.AddMinutes(10));
+            }
+
+            void UploadObjectSucceeds_DownloadObjectFails()
+            {
+                string objectName = IdGenerator.FromGuid();
+                var objectData = TestHelpers.GenerateData(512);
+                client.UploadObject(bucketName, objectName, "", objectData);
+                Assert.Throws<GoogleApiException>(() => client.DownloadObject(bucketName, objectName, new MemoryStream()));
+            }
```

Comment:
Does this test using granular permissions when a service account has permission to upload but not download? Not sure what changed to manage the permissions though.

Reply:
I don't know the details, to be honest - but I believe this is the expected (by the Storage team :) behaviour for the case where BPO is enabled after bucket creation.